### PR TITLE
Add Basenames decoder

### LINKED ISSUE #8602: Decode base names
Example transaction https://basescan.org/tx/0x2e910a97dfd95d7b6e975a822c0289f3dd4a6fdf400c6c677502e5f6eebe8d76

Subdomains of base.eth https://www.base.org/names

## Task

Decode name registrations and extensions. Should be similar to ENS

### PR DESCRIPTION
**Pull request by nicholasyoder**
Closes #8602

### CODE REVIEW

File: rotkehlchen/chain/evm/decoding/ens/decoder.py [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+from rotkehlchen.chain.ethereum.modules.ens.constants import (
+    ADDR_CHANGED,
+    CONTENT_HASH_CHANGED,
+    NEW_OWNER,
+    NEW_RESOLVER,
+    TEXT_CHANGED_KEY_AND_VALUE,
+    TEXT_CHANGED_KEY_AND_VALUE_ABI,
+    TEXT_CHANGED_KEY_ONLY,
+    TEXT_CHANGED_KEY_ONLY_ABI,
+)
```
Comment (prettyirrelevant) [inline rotkehlchen/chain/evm/decoding/ens/decoder.py:19]: these constants should be at `evm/decoding/ens/constants.py` or what do you think?

**Review by prettyirrelevant — COMMENTED** [review]
thanks @nicholasyoder. left some comments

File: rotkehlchen/chain/base/modules/basenames/decoder.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+
+        fullname = f'{decoded_data[0]}.base.eth'
+        expires = decoded_data[1]
+        spend_event = receive_event = None
+        for event in context.decoded_events:
+            if event.event_type == HistoryEventType.SPEND and event.asset == A_ETH and event.address == context.tx_log.address:  # noqa: E501
+                event.event_type = HistoryEventType.TRADE
+                event.event_subtype = HistoryEventSubType.SPEND
+                event.counterparty = CPT_BASENAMES
+                event.notes = f'Register ENS name {fullname} for {event.balance.amount} ETH until {self.timestamp_to_date(expires)}'  # noqa: E501
```
[the remark this answers is not in the document]
Comment (nicholasyoder) [reply rotkehlchen/chain/base/modules/basenames/decoder.py]: Yes, should be Basename, although to make it consistent with how its handled in the [EnsCommonDecoder](https://github.com/rotki/rotki/pull/8891/files#diff-c6f252084d3dbca1d7d0fc2df40f41de19be2a26e2f52203cbaa8cad6c60ab26R113), I think it should be `Basenames name`.

File: rotkehlchen/chain/base/modules/basenames/decoder.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+
+        fullname = f'{decoded_data[0]}.base.eth'
+        expires = decoded_data[1]
+        spend_event = receive_event = None
+        for event in context.decoded_events:
+            if event.event_type == HistoryEventType.SPEND and event.asset == A_ETH and event.address == context.tx_log.address:  # noqa: E501
+                event.event_type = HistoryEventType.TRADE
+                event.event_subtype = HistoryEventSubType.SPEND
+                event.counterparty = CPT_BASENAMES
+                event.notes = f'Register Basenames name {fullname} for {event.balance.amount} ETH until {self.timestamp_to_date(expires)}'  # noqa: E501
```
Comment (prettyirrelevant) [inline rotkehlchen/chain/base/modules/basenames/decoder.py:177]:
```suggestion
                event.notes = f'Register {self.display_name} name {fullname} for {event.balance.amount} ETH until {self.timestamp_to_date(expires)}'  # noqa: E501
```
might as well utilize the attribute.

File: rotkehlchen/chain/base/modules/basenames/decoder.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+            # Create spend event with zero balance for discounted registrations
+            spend_event = self.base.make_event_from_transaction(
+                transaction=context.transaction,
+                tx_log=context.tx_log,
+                event_type=HistoryEventType.TRADE,
+                event_subtype=HistoryEventSubType.SPEND,
+                asset=A_ETH,
+                balance=Balance(amount=ZERO),
+                location_label=bytes_to_address(context.tx_log.topics[2]),
+                notes=f'Register Basenames name {fullname} for 0 ETH until {self.timestamp_to_date(expires)}',  # noqa: E501
```
Comment (prettyirrelevant) [inline rotkehlchen/chain/base/modules/basenames/decoder.py:195]:
```suggestion
                notes=f'Register {self.display_name} name {fullname} for 0 ETH until {self.timestamp_to_date(expires)}',  # noqa: E501
```
same as above

File: rotkehlchen/chain/base/modules/basenames/decoder.py [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+            with GlobalDBHandler().conn.write_ctx() as write_cursor:
+                globaldb_set_unique_cache_value(
+                    write_cursor=write_cursor,
+                    key_parts=(CacheType.ENS_LABELHASH, label_hash),
+                    value=(name := response_data['name']),
+                )
+        except (TypeError, KeyError):
+            log.error(
+                f'Failed to retrieve name for labelhash {label_hash} '
+                f'due to unexpected response from Basenames api: {response_data}',
```
Comment (prettyirrelevant) [inline rotkehlchen/chain/base/modules/basenames/decoder.py:150]: you probably need to change the msg here.

File: rotkehlchen/chain/base/modules/basenames/decoder.py [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+                    cursor=cursor,
+                    key_parts=(CacheType.ENS_NAMEHASH, node.hex()),
+            )) is not None:
+                return name_to_show
+
+        if (name_to_show := self.evm_inquirer.contracts.contract(BASENAMES_L2_RESOLVER).call(
+            node_inquirer=self.evm_inquirer,
+            method_name='name',
+            arguments=[node],
+        )):
```
Comment (prettyirrelevant) [inline rotkehlchen/chain/base/modules/basenames/decoder.py:97]: perhaps handle exceptions thrown by this call

**Review by prettyirrelevant — COMMENTED** [review]
thanks, small comments.

File: rotkehlchen/chain/base/modules/basenames/decoder.py [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+
+        fullname = f'{decoded_data[0]}.base.eth'
+        expires = decoded_data[1]
+        spend_event = receive_event = None
+        for event in context.decoded_events:
+            if event.event_type == HistoryEventType.SPEND and event.asset == A_ETH and event.address == context.tx_log.address:  # noqa: E501
+                event.event_type = HistoryEventType.TRADE
+                event.event_subtype = HistoryEventSubType.SPEND
+                event.counterparty = CPT_BASENAMES
+                event.notes = f'Register {self.display_name} name {fullname} for {event.balance.amount} ETH until {self.timestamp_to_date(expires)}'  # noqa: E501
```
[the remark this answers is not in the document]
Comment (yabirgb) [reply rotkehlchen/chain/base/modules/basenames/decoder.py:181]: I paid for my name. I didn´t use the coin wallet but the web app

File: rotkehlchen/chain/base/modules/basenames/decoder.py [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+            )) is not None:
+                return name_to_show
+
+        try:
+            if not (name_to_show := self.evm_inquirer.contracts.contract(BASENAMES_L2_RESOLVER).call(  # noqa: E501
+                node_inquirer=self.evm_inquirer,
+                method_name='name',
+                arguments=[node],
+            )):
+                raise RemoteError('blank name returned from resolver contract.')
```
[the remark this answers is not in the document]
Comment (nicholasyoder) [reply rotkehlchen/chain/base/modules/basenames/decoder.py:100]: Just below this its catching any RemoteErrors and logging them with the node hex included. This way any remote errors from the contract call are also handled.

File: rotkehlchen/chain/base/modules/basenames/decoder.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (nicholasyoder) [reply rotkehlchen/chain/base/modules/basenames/decoder.py:180]: Leaving this as it is, as mentioned in [discord](https://discord.com/channels/657906918408585217/1080783409640648765/1306212779933433917)

File: rotkehlchen/chain/evm/decoding/ens/decoder.py [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+        return DEFAULT_DECODING_OUTPUT
+
+    def _decode_ens_public_resolver_content_hash(self, context: DecoderContext) -> DecodingOutput:
+        """Decode an event that modifies a content hash for the public ENS resolver"""
+        node = context.tx_log.topics[1]  # node is a hash of the name used by ens internals
+        contract = self.evm_inquirer.contracts.contract_by_address(address=context.tx_log.address)
+        if contract is None:
+            self.msg_aggregator.add_error(
+                f'Failed to find {self.display_name} public resolver contract with address '
+                f'{context.tx_log.address} for {context.transaction.tx_hash.hex()}. '
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/decoding/ens/decoder.py:186]: need to also specify chain name in this error now

File: rotkehlchen/chain/evm/decoding/ens/decoder.py [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+
+        try:
+            codec = content_hash.get_codec(new_hash)
+            value_hash = content_hash.decode(new_hash)
+            value = f'{codec}://{value_hash}'
+        except (TypeError, KeyError, ValueError) as e:
+            msg = str(e)
+            if isinstance(e, KeyError):
+                msg = f'Inability to find key {msg}'
+            log.error(f'Failed to decode content hash {new_hash} in {context.transaction.tx_hash.hex()} due to {msg}')  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/decoding/ens/decoder.py:204]: need to specify chain name in this error now

File: rotkehlchen/chain/evm/decoding/ens/decoder.py [hunk 11]
```diff
@@ -0,0 +1,321 @@
...
+        if context.tx_log.topics[0] not in (TEXT_CHANGED_KEY_ONLY, TEXT_CHANGED_KEY_AND_VALUE):
+            return DEFAULT_DECODING_OUTPUT
+
+        try:
+            _, decoded_data = decode_event_data_abi_str(
+                context.tx_log,
+                TEXT_CHANGED_KEY_ONLY_ABI if context.tx_log.topics[0] == TEXT_CHANGED_KEY_ONLY else TEXT_CHANGED_KEY_AND_VALUE_ABI,  # noqa: E501
+            )
+        except DeserializationError as e:
+            log.error(f'Failed to decode {self.display_name} set-text event in {context.transaction.tx_hash.hex()} due to {e!s}')  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/decoding/ens/decoder.py:272]: need to specify chain name in this error now.

File: rotkehlchen/chain/evm/decoding/ens/decoder.py [hunk 12]
```diff
@@ -0,0 +1,321 @@
...
+if TYPE_CHECKING:
+    from rotkehlchen.chain.evm.decoding.base import BaseDecoderTools
+    from rotkehlchen.chain.evm.node_inquirer import EvmNodeInquirer
+    from rotkehlchen.user_messages import MessagesAggregator
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+
+class EnsCommonDecoder(DecoderInterface, CustomizableDateMixin):
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/decoding/ens/decoder.py:45]: This should be an abstract class.

File: rotkehlchen/chain/evm/decoding/ens/decoder.py [hunk 13]
```diff
@@ -0,0 +1,321 @@
...
+            asset=A_ETH,
+            balance=Balance(),
+            location_label=context.transaction.from_address,
+            notes=notes,
+            counterparty=self.counterparty,
+            address=context.transaction.to_address,
+        ))
+        return DEFAULT_DECODING_OUTPUT
+
+    def _maybe_get_labelhash_name(
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/decoding/ens/decoder.py:297]: This should be an abstract method

File: rotkehlchen/chain/evm/decoding/ens/decoder.py [hunk 14]
```diff
@@ -0,0 +1,321 @@
...
+            context: DecoderContext,
+            label_hash: str,
+            node: bytes | None = None,
+    ) -> str | None:
+        """
+        Subclasses implement this to retrieve the name for a labelhash
+        """
+        return None
+
+    def _get_name_to_show(self, node: bytes, context: DecoderContext) -> str | None:
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/decoding/ens/decoder.py:308]: This should be an abstract method

File: rotkehlchen/tests/unit/decoders/test_basenames.py [hunk 15]
```diff
@@ -0,0 +1,468 @@
...
+            address=BASENAMES_REGISTRAR_CONTROLLER,
+        ), EvmEvent(
+            tx_hash=tx_hash,
+            sequence_index=186,
+            timestamp=timestamp,
+            location=Location.BASE,
+            event_type=HistoryEventType.INFORMATIONAL,
+            event_subtype=HistoryEventSubType.NONE,
+            asset=A_ETH,
+            balance=Balance(amount=ZERO),
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/decoders/test_basenames.py:85]: please remember in all your code. this is unnecessary

```suggestion
            balance=Balance(),
```

File: rotkehlchen/tests/unit/decoders/test_basenames.py [hunk 16]
```diff
@@ -0,0 +1,468 @@
...
+            counterparty=CPT_GAS,
+        ), EvmEvent(
+            tx_hash=tx_hash,
+            sequence_index=402,
+            timestamp=timestamp,
+            location=Location.BASE,
+            event_type=HistoryEventType.INFORMATIONAL,
+            event_subtype=HistoryEventSubType.NONE,
+            asset=A_ETH,
+            balance=Balance(amount=ZERO),
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/decoders/test_basenames.py:178]: same here and generally in all your code

File: rotkehlchen/chain/base/modules/basenames/decoder.py [hunk 17]
```diff
@@ -0,0 +1,224 @@
...
+            return None
+
+        try:
+            with GlobalDBHandler().conn.write_ctx() as write_cursor:
+                globaldb_set_unique_cache_value(
+                    write_cursor=write_cursor,
+                    key_parts=(CacheType.ENS_LABELHASH, label_hash),
+                    value=(name := response_data['name']),
+                )
+        except (TypeError, KeyError):
```
Comment (LefterisJP) [inline rotkehlchen/chain/base/modules/basenames/decoder.py:149]: So this is not giving enough information in the logs.

If what you want to check here is that `response_data` is a dict then you should explicitly check for that. There is no way to know when you read the code why the type error is checked for. Check for it above.

Also then this just leaves key error in which case you should simply mention it's a missing attribute (and which one it is) from the response.

File: rotkehlchen/chain/evm/decoding/ens/decoder.py [hunk 18]
```diff
@@ -0,0 +1,311 @@
...
+if TYPE_CHECKING:
+    from rotkehlchen.chain.evm.decoding.base import BaseDecoderTools
+    from rotkehlchen.chain.evm.node_inquirer import EvmNodeInquirer
+    from rotkehlchen.user_messages import MessagesAggregator
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+
+class EnsCommonDecoder(DecoderInterface, CustomizableDateMixin):
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/decoding/ens/decoder.py:44]: you did not make it ann abstract class. It needs to inherit from ABC. Check the rest of the code. Check what an abstract class is and don't resolve comments in the PR without doing them or if you don't understand them then ask.

Comment (nicholasyoder) [reply rotkehlchen/chain/evm/decoding/ens/decoder.py]: Now inherits ABC, but see question in [discord](https://discord.com/channels/657906918408585217/1080783409640648765/1306333798644060281).

Comment (LefterisJP) [reply rotkehlchen/chain/evm/decoding/ens/decoder.py]: it is required as explained in discord.

**Review by LefterisJP — APPROVED** [review]
LGTM

**Pull request merged**